session/sync_upgrade.go

```diff
@@ -0,0 +1,139 @@
+		}
+		time.Sleep(interval)
+	}
+
+	logutil.BgLogger().Info("update global state to upgrading", zap.String("category", "upgrading"), zap.String("state", syncer.StateUpgrading))
+	return nil
+}
+
+// SyncNormalRunning syncs normal state to etcd.
+func SyncNormalRunning(s sessionctx.Context) error {
+	failpoint.Inject("mockResumeAllJobsFailed", func(val failpoint.Value) {
+		if val.(bool) {
+			dom := domain.GetDomain(s)
+			//nolint: errcheck
+			dom.DDL().StateSyncer().UpdateGlobalState(context.Background(), syncer.NewStateInfo(syncer.StateNormalRunning))
+			failpoint.Return(nil)
+		}
+	})
+
+	jobErrs, err := ddl.ResumeAllJobsBySystem(s)
+	if err != nil {
+		logutil.BgLogger().Warn("resume all paused jobs failed", zap.String("category", "upgrading"), zap.Error(err))
+	}
+	for _, e := range jobErrs {
+		logutil.BgLogger().Warn("resume the job failed ", zap.String("category", "upgrading"), zap.Error(e))
+	}
+
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancelFunc()
+	dom := domain.GetDomain(s)
+	err = dom.DDL().StateSyncer().UpdateGlobalState(ctx, syncer.NewStateInfo(syncer.StateNormalRunning))
+	if err != nil {
+		logutil.BgLogger().Error("update global state to normal failed", zap.String("category", "upgrading"), zap.Error(err))
+		return err
+	}
+	logutil.BgLogger().Info("update global state to normal running finished", zap.String("category", "upgrading"))
+	return nil
+}
+
+// IsUpgradingClusterState checks whether the global state is upgrading.
+func IsUpgradingClusterState(s sessionctx.Context) (bool, error) {
+	dom := domain.GetDomain(s)
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancelFunc()
+	stateInfo, err := dom.DDL().StateSyncer().GetGlobalState(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	return stateInfo.State == syncer.StateUpgrading, nil
+}
+
+func checkOrSyncUpgrade(s Session, ver int64) {
+	if ver < SupportUpgradeHTTPOpVer {
+		terror.MustNil(SyncUpgradeState(s, time.Duration(internalSQLTimeout)*time.Second))
+		return
+	}
+	isUpgradingClusterStateWithRetry(s, ver, currentBootstrapVersion, time.Duration(internalSQLTimeout)*time.Second)
+}
+
+func isUpgradingClusterStateWithRetry(s sessionctx.Context, oldVer, newVer int64, timeout time.Duration) {
+	interval := 200 * time.Millisecond
+	retryTimes := int(timeout / interval)
+	for i := 0; i < retryTimes; i++ {
+		isUpgrading, err := IsUpgradingClusterState(s)
+		if err == nil {
+			if isUpgrading {
+				break
+			}
+			logutil.BgLogger().Fatal("global state isn't upgrading, please send a request to start the upgrade first",
+				zap.String("category", "upgrading"), zap.Error(err))
+		}
+
+		if i == retryTimes-1 {
+			logutil.BgLogger().Fatal("get global state failed", zap.String("category", "upgrading"), zap.Error(err))
+		}
+		if i%10 == 0 {
+			logutil.BgLogger().Warn("get global state failed", zap.String("category", "upgrading"), zap.Error(err))
+		}
+		time.Sleep(interval)
```

Comment:
time.Sleep should add the random time. it can avoid conflicting with other service to happen the timeout.
```
		time.Sleep(interval) + rand.Int() 
```